src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs

```diff
@@ -45,464 +46,505 @@ public static IEnumerable<object[]> Match_Basic_TestData_LazyLoops()
-        public async Task Match_NetCore(RegexEngine engine, string pattern, string input, RegexOptions options, int beginning, int length, bool expectedSuccess, string expectedValue)
-        {
-            await Match(engine, pattern, input, options, beginning, length, expectedSuccess, expectedValue);
+                // Not naming pattern slots at all: "^(cat|chat)"
+                yield return new object[] { engine, "^(cat|chat)", "cats are bad", RegexOptions.None, 0, 12, true, "cat" };
+
+                yield return new object[] { engine, "abc", "abc", RegexOptions.None, 0, 3, true, "abc" };
+                yield return new object[] { engine, "abc", "aBc", RegexOptions.None, 0, 3, false, string.Empty };
+                yield return new object[] { engine, "abc", "aBc", RegexOptions.IgnoreCase, 0, 3, true, "aBc" };
+                yield return new object[] { engine, @"abc.*def", "abcghiDEF", RegexOptions.IgnoreCase, 0, 9, true, "abcghiDEF" };
+
+                // Using *, +, ?, {}: Actual - "a+\\.?b*\\.?c{2}"
+                yield return new object[] { engine, @"a+\.?b*\.+c{2}", "ab.cc", RegexOptions.None, 0, 5, true, "ab.cc" };
+                yield return new object[] { engine, @"[^a]+\.[^z]+", "zzzzz", RegexOptions.None, 0, 5, false, string.Empty };
+
+                // IgnoreCase
+                yield return new object[] { engine, "AAA", "aaabbb", RegexOptions.IgnoreCase, 0, 6, true, "aaa" };
+                yield return new object[] { engine, @"\p{Lu}", "1bc", RegexOptions.IgnoreCase, 0, 3, true, "b" };
+                yield return new object[] { engine, @"\p{Ll}", "1bc", RegexOptions.IgnoreCase, 0, 3, true, "b" };
+                yield return new object[] { engine, @"\p{Lt}", "1bc", RegexOptions.IgnoreCase, 0, 3, true, "b" };
+                yield return new object[] { engine, @"\p{Lo}", "1bc", RegexOptions.IgnoreCase, 0, 3, false, string.Empty };
+
+                // "\D+"
+                yield return new object[] { engine, @"\D+", "12321", RegexOptions.None, 0, 5, false, string.Empty };
+
+                // Groups
+                yield return new object[] { engine, "(?<first_name>\\S+)\\s(?<last_name>\\S+)", "David Bau", RegexOptions.None, 0, 9, true, "David Bau" };
+
+                // "^b"
+                yield return new object[] { engine, "^b", "abc", RegexOptions.None, 0, 3, false, string.Empty };
+
+                // Trim leading and trailing whitespace
+                yield return new object[] { engine, @"\s*(.*?)\s*$", " Hello World ", RegexOptions.None, 0, 13, true, " Hello World " };
+
+                if (!RegexHelpers.IsNonBacktracking(engine))
+                {
+                    // TODO-NONBACKTRACKING: This test fails by successfully matching.
+                    // < in group
+                    yield return new object[] { engine, @"(?<cat>cat)\w+(?<dog-0>dog)", "cat_Hello_World_dog", RegexOptions.None, 0, 19, false, string.Empty };
```

Comment:
Seems something is amiss here in the NonBacktracking implementation... the test is expecting matching to fail but NonBacktracking actually produces a match.

Reply:
Oooh, this is **balancing group**, didn't even know this existed until now. **Indeed**, we need to throw `NotSupportedException` for this case, it is a nonregular feature (can be used e.g. to match some context-free expressions like balanced parenthesis)

Reply:
fixed in #1613